rskj-core/src/test/java/co/rsk/peg/federation/FederationChangeIT.java

```diff
@@ -414,8 +552,130 @@ private void assertNewAndOldFederationsHaveExpectedAddress(
+        var lastRetiredFederationP2SHScriptOptional = 
+            federationStorageProvider.getLastRetiredFederationP2SHScript(activations);
+        assertTrue(lastRetiredFederationP2SHScriptOptional.isPresent());
+        Script lastRetiredFederationP2SHScript = lastRetiredFederationP2SHScriptOptional.get();
+
+        if (activations.isActive(ConsensusRule.RSKIP377)){
+            if (federationType == FederationType.NON_STANDARD_ERP
+                || federationType == FederationType.P2SH_ERP) {
+                assertNotEquals(lastRetiredFederationP2SHScript, originalFederation.getP2SHScript());
+            }
+            assertEquals(lastRetiredFederationP2SHScript, getFederationDefaultP2SHScript(originalFederation));
+        } else {
+            if (federationType == FederationType.NON_STANDARD_ERP
+                || federationType == FederationType.P2SH_ERP) {
+                assertEquals(lastRetiredFederationP2SHScript, originalFederation.getP2SHScript());
+                assertNotEquals(lastRetiredFederationP2SHScript, getFederationDefaultP2SHScript(originalFederation));
+            } else {
+                assertEquals(lastRetiredFederationP2SHScript, originalFederation.getP2SHScript());
+                assertEquals(lastRetiredFederationP2SHScript, getFederationDefaultP2SHScript(originalFederation));
+            }
+        }
+    }
+   
+    private void verifySigHashes(ActivationConfig.ForBlock activations) throws Exception {
+        var pegoutsTxs = bridgeStorageProvider.getPegoutsWaitingForConfirmations()
+            .getEntries().stream()
+            .map(Entry::getBtcTransaction)
+            .toList();
+
+        pegoutsTxs.forEach(
+            pegoutTx -> verifyPegoutTxSigHashIndex(pegoutTx, activations));
+    }
+
+    private void verifyPegoutTransactionCreatedEventWasEmitted(ActivationConfig.ForBlock activations) throws Exception {
+        var pegoutsTxs = bridgeStorageProvider.getPegoutsWaitingForConfirmations()
+            .getEntries().stream()
+            .map(Entry::getBtcTransaction)
+            .toList();
+
+        if (activations.isActive(ConsensusRule.RSKIP428)) {
+            pegoutsTxs.forEach(this::verifyPegoutTransactionCreatedEvent);
+        } else {
+            verify(bridgeEventLogger, never()).logPegoutTransactionCreated(any(), any());
+        }
+    }
+    
+    private void verifyPegouts(ActivationConfig.ForBlock activations) throws Exception {
+        var activeFederation = federationStorageProvider.getNewFederation(
+            BRIDGE_CONSTANTS.getFederationConstants(), activations);
+        var retiringFederation = federationStorageProvider.getOldFederation(
+            BRIDGE_CONSTANTS.getFederationConstants(), activations);
+
+        for (PegoutsWaitingForConfirmations.Entry pegoutEntry : bridgeStorageProvider.getPegoutsWaitingForConfirmations().getEntries()) {
```

Comment:
Not sure about this. I think we should assert against constant values. I mean, for example, based on the initial conditions of the test, we should know the number of migration pegouts that should be created and assert that that number of pegouts is precisely what gets created. Then, check each one was created as expected. Wdyt?

Reply:
The UTXOs for the original federation are randomly generated. So the size may vary per test. It was already done in the legacy tests here: https://github.com/rsksmart/rskj/blob/feature/powpeg_validation_protocol-phase4/rskj-core/src/test/java/co/rsk/peg/federation/PowpegMigrationTest.java

Reply:
We've been removing/getting away from random behavior from tests. I think, we should discuss if we should keep this being random or make it deterministic.